keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1550 @@
+  resource types in the future.
+- Make QoS-class resources opqaue (as possible) to the CRI client
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Interface or mechanism for configuring the QoS-class resources (responsibility of
+  the container runtime).
+- Enumerating possible (QoS-class) resource types or their detailed behavior
+- API changes to support updating Pod-level (sandbox-level) QoS-class resource
+  assignment of running pods (will be addressed in a separate KEP)
+- Resource status/capacity (will be addressed in a separate KEP)
+- Discovery of the QoS-class resources (will be addressed in a separate KEP)
+- Access control (will be addressed in a separate KEP)
+
+## Implementation phases
+
+This proposal splits the full implementation of QoS-class resources into
+multiple phases, building functionality gradually, step-by-step. The goal is to
+make the discussions more focused and easier. We may also learn on the way,
+insights from earlier phases affecting design choises made in the later phases,
+hopefully resulting in a better overall end result. However, we also outline
+all the future steps to not lose the overall big picture.
+
+In the current design QoS-class resources are designed to be opaque to the CRI
+client in the sense that the container runtime takes care of configuration and
+control of the resources and the classes within.
+
+### Phase 1
+
+This KEP (the [Proposal](#proposal)) implements the first phase. The goal is to
+enable a bare minimum for users to leverage QoS-class resources and start
+experimenting with them in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates
+- implement pod annotations as an initial user interface
+- introduce a feature gate for enabling QoS-class resource support in kubelet
+
+### Future work
+
+This section sheds light on the end goal of this work in order to better
+evaluate this KEP in a broader context. What a fully working solution would
+consists of and what the (next) steps to accomplish that would be. These topics
+are currently out of the scope of this KEP and were listed under
+[Non-goals](#non-goals).
+
+#### Pod Spec
+
+Replace pod annotations with proper user interface via the Pod spec. Below, one
+possible option is presented.
+
+Introduce a new field (e.g. class) into ResourceRequirements of Container.
+
+```diff
+// ResourceRequirements describes the compute resource requirements.
+type ResourceRequirements struct {
+     // Limits describes the maximum amount of compute resources allowed.
+     Limits ResourceList `json:"limits,omitempty"
+     // Requests describes the minimum amount of compute resources required.
+     Requests ResourceList `json:"requests,omitempty"
++    // Classes specifies the QoS-class resources that the container should be assigned
++    Classes map[ClassResourceName]string
```

Comment:
can containers in the same pod have different classes? do we need to specify a `Resources` field in the PodSpec if the answer is no?

Reply:
Different containers definitely can have different classes (think about e.g. a high priority main workload plus a log handler).

The container-level and pod-level (qos-class) resources are now totally independent/orthogonal concepts in this KEP (I tried to capture that in the last paragraph of the "Motivation" section). E.g. in case of RDT there cannot be such a thing as a sandbox-level class because of hw-limitations (it's a flat one-level hierarchy).

Makes sense?